PhysicsTools/PatAlgos/python/jetCollectionTools.py

```diff
@@ -0,0 +1,621 @@
+
+    if not recoJetInfo.skipUserData:
+      #
+      jercVar = "jercVars{}".format(tagName)
+      if jercVar in self.main:
+        raise ValueError("Step '%s' already implemented" % jercVar)
+      setattr(proc, jercVar, proc.jercVars.clone(srcJet = cms.InputTag(selJet)))
+      currentTasks.append(jercVar)
+      #
+      looseJetId = "looseJetId{}".format(tagName)
+      if looseJetId in self.main:
+        raise ValueError("Step '%s' already implemented" % looseJetId)
+      setattr(proc, looseJetId, proc.looseJetId.clone(src = cms.InputTag(selJet)))
+      #
+      tightJetId = "tightJetId{}".format(tagName)
+      if tightJetId in self.main:
+        raise ValueError("Step '%s' already implemented" % tightJetId)
+      setattr(proc, tightJetId, proc.tightJetId.clone(src = cms.InputTag(selJet)))
+      currentTasks.append(tightJetId)
+      #
+      tightJetIdLepVeto = "tightJetIdLepVeto{}".format(tagName)
+      if tightJetIdLepVeto in self.main:
+        raise ValueError("Step '%s' already implemented" % tightJetIdLepVeto)
+      setattr(proc, tightJetIdLepVeto, proc.tightJetIdLepVeto.clone(src = cms.InputTag(selJet)))
+      currentTasks.append(tightJetIdLepVeto)
+      #
+      selectedPatJetsWithUserData = "{}WithUserData".format(selJet)
+      if selectedPatJetsWithUserData in self.main:
+        raise ValueError("Step '%s' already implemented" % selectedPatJetsWithUserData)
+      setattr(proc, selectedPatJetsWithUserData,
+        cms.EDProducer("PATJetUserDataEmbedder",
+          src = cms.InputTag(selJet),
+          userFloats = cms.PSet(
+            jercCHPUF = cms.InputTag("{}:chargedHadronPUEnergyFraction".format(jercVar)),
+            jercCHF   = cms.InputTag("{}:chargedHadronCHSEnergyFraction".format(jercVar)),
+          ),
+          userInts = cms.PSet(
+            tightId        = cms.InputTag(tightJetId),
+            tightIdLepVeto = cms.InputTag(tightJetIdLepVeto),
+          ),
+        )
+      )
+      for modifier in run2_miniAOD_80XLegacy, run2_nanoAOD_94X2016:
+        selectedPatJetsWithUserDataObj = getattr(proc, selectedPatJetsWithUserData)
+        modifier.toModify(selectedPatJetsWithUserDataObj.userInts,
+          looseId        = cms.InputTag(looseJetId),
+          tightIdLepVeto = None,
+        )
+      currentTasks.append(selectedPatJetsWithUserData)
+    else:
+      selectedPatJetsWithUserData = "selectedPatJets{}".format(tagName)
+
+    # Not sure why we can't re-use patJetCorrFactors* created by addJetCollection() (even cloning doesn't work)
+    # Let's just create our own
+    jetCorrFactors = "jetCorrFactors{}".format(tagName)
+    if jetCorrFactors in self.main:
+      raise ValueError("Step '%s' already implemented" % jetCorrFactors)
+    setattr(proc, jetCorrFactors, patJetCorrFactors.clone(
+        src             = selectedPatJetsWithUserData,
+        levels          = cms.vstring(JETCorrLevels),
+        primaryVertices = cms.InputTag(self.pvLabel),
```

Comment:
please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.